include/seqan3/contrib/parallel/reader_writer_latch.hpp

```diff
@@ -0,0 +1,323 @@
+     * count set to 0.
+     */
+    template <typename concurrent_t>
+        requires requires { std::declval<concurrent_t>().close(); }  // requires closable queue.
+    reader_writer_latch(writer_count const wcount, concurrent_t & queue) :
+        reader_writer_latch{wcount, reader_count{0}, queue}
+    {}
+
+    /*!\brief Constructs the reader_writer_latch with the reader count and the associated data structure.
+     * \tparam concurrent_t The type of the concurrent data structure. Must have a member function `close`.
+     * \param rcount The expected number of consumer threads.
+     * \param ds     The concurrent data structure.
+     *
+     * \details
+     *
+     * Delegates to reader_writer_latch::reader_writer_latch(reader_count, writer_count, concurrent_t) with writer
+     * count set to 0.
+     */
+    template <typename concurrent_t>
+        requires requires { std::declval<concurrent_t>().close(); }  // requires closable queue.
+    reader_writer_latch(reader_count const rcount, concurrent_t & queue) :
+        reader_writer_latch{writer_count{0}, rcount, queue}
+    {}
+    //!\}
+
+    /*!\brief Atomically decrements writer counter by one and blocks the calling thread.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for producer threads and blocks the calling thread until all participating
+     * threads have arrived. If all participating producer threads arrived at the synchronisation point, one of the
+     * threads will invoke the completion function.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive_and_wait() noexcept
+    {
+        writer_latch.arrive_and_wait();
+
+        std::call_once(flag, completion_fn);
+    }
+
+    /*!\brief Atomically decrements writer counter by one.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for producer threads without blocking the calling thread.
+     * If all participating producer threads arrived at the synchronisation point, one of the
+     * threads will invoke the completion function.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive() noexcept
+    {
+        writer_latch.arrive();
+
+        if (writer_latch.try_wait())
+            std::call_once(flag, completion_fn);
+    }
+
+    /*!\brief Atomically decrements writer counter by one and blocks the calling thread.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for consumer threads and blocks the calling thread until all participating
+     * threads have arrived.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void reader_arrive_and_wait() noexcept
+    {
+        reader_latch.arrive_and_wait();
```

Comment:
How are the readers synchronised with the writers?

So if `writer_arrive_and_wait` was successful and closes the data structure, but there are still readers calling `arrive_and_wait`.

WARNING: This may be a stupid question.

Reply:
The readers are participating in a distinct synchronisation point. So if the writers are finished, the ds is closed, such that the readers can finish until the ds is empty or some other event removes them from the computation. Thus with this latch it is possible to avoid destructing a queue, simply by putting the latch after the queue in the member list. Thus on destruction of the wrapping data structure first the latch and then the queue is destructed, but only if all threads have finished there work.

Reply:
yes